[AMDGPU] Asynchronous operations with explicit marks and waits

ssahasra opened the pull request:
Asynchronous operations are memory transfers (usually between the global memory
and LDS) that are completed independently at an unspecified scope. A thread that
requests one or more asynchronous transfers can use *async marks* to track their
completion. The thread waits for each mark to be *completed*, which indicates
that requests initiated in program order before this mark have also completed.

This patch introduces the following builtins/intrinsics:
- async loads from global/buffer memory to LDS
- asyncmark and wait operations to track their completion
- Support in SIInsertWaitcnts to correctly track marks and waits

For now, we implement asyncmark/wait operations on pre-GFX12 architectures that
support "LDS DMA" operations. Future work will extend support to GFX12Plus
architectures that support "true" async operations.

Co-authored-by: Ryan Mitchell <ryan.mitchell@amd.com>

Fixes: SWDEV-521121

jayfoad reviewed (commented):
Mostly cosmetic things.

RyanRio reviewed (commented):
Functionally good, maybe we could note some TODOs here in regards to marking what regs are part of an async operation and how that interplays with waitcnts generated/deemed necessary for synchronous operations.

ssahasra commented on llvm/test/CodeGen/AMDGPU/async-mark-pregfx12.ll:
```diff
@@ -0,0 +1,898 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 6
+; RUN: llc -march=amdgcn -mcpu=gfx900  < %s | FileCheck %s -check-prefixes=GFX900
+; RUN: llc -march=amdgcn -mcpu=gfx942  < %s | FileCheck %s -check-prefixes=GFX942
```
Is this whitespace really a bad thing? Having it aligns all the other keywords making it easier to spot the difference in the RUN lines.

ssahasra commented on llvm/lib/Target/AMDGPU/SOPInstructions.td:
```diff
@@ -1688,11 +1688,22 @@ let SubtargetPredicate = HasWaitXcnt in {
 }
 
+// FIXME: Until support for GFX12Plus is ready, restrict these to only preGFX12
+let SubtargetPredicate = isGFX9GFX10GFX11 in {
```
Fixed. Had to introduce the predicate and rename the existing Feature for consistency.

ssahasra commented:
> Functionally good, maybe we could note some TODOs here in regards to marking what regs are part of an async operation and how that interplays with waitcnts generated/deemed necessary for synchronous operations.

Right most of those TODOs are port of the next piece of work. For now, I have added a comment saying that LOAD_CNT is not the only counter, and other counters will get involved with future work.

nhaehnle commented on llvm/test/CodeGen/AMDGPU/asyncmark-pregfx12.ll:
```diff
@@ -0,0 +1,898 @@
+  call void @llvm.amdgcn.wave.barrier()
+  call void @llvm.amdgcn.global.load.lds(ptr addrspace(1) %bar_gep21, ptr addrspace(3) %lds_gep21, i32 4, i32 0, i32 u0x21)
+  call void @llvm.amdgcn.wave.barrier()
+  call void @llvm.amdgcn.asyncmark()
```
What's the logic behind having wave.barriers here (and elsewhere)? They are unexpected to me.

(I briefly wondered about `convergent` in all this, but IMO the way to go is to conceptually make the asyncmarks a thread-level concept in LLVM IR. They are turned into wave-level concepts in assembly, but the worst that can happen is more conservative waits.)

nhaehnle reviewed (commented):
Mostly LGTM.

I suspect there's an infinite loop due to how `mergeAsyncMarkers` works when the IR contains a loop that has an `asyncmark` intrinsic but no `wait.asyncmark`. What prevents the AsyncMarkers array from just growing indefinitely?

It would generally be nice to have some additional tests where the asyncmarkers in control flow aren't balanced. BTW, I think it'd be fine to have tests that are a bit more bare-bone, e.g. always using the same pointer arguments for the load.lds. That would result in less "noise" in the  assembly.

krzysz00 reviewed (commented):
This is quite nice and I think will be helpful going forward.

@Hardcode84 @qedawkins @antiagainst @lialan @kuhar @bjacob as the MLIR-level folks who may be interested in using this and so are getting tagged (mostly as "hey, check out the cool new thing" with a side of "please say something now if you see any glaring red flags with this API")

Hardcode84 commented:
Very nice. It'd be probably useful to have this also for LDS stores from registers at some point.

nhaehnle commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -653,6 +677,11 @@ class WaitcntBrackets {
+  unsigned getClampedWait(InstCounterType T, unsigned ScoreToWait) const {
+    return std::min(getScoreUB(T) - ScoreToWait,
+                    Context->getWaitCountMax(T) - 1);
+  }
```
Perhaps this check could just be added to `addWait` instead?

ssahasra replied:
That function is not critical to this change. I just reverted it for now.

Pierre-vh commented on llvm/docs/AMDGPUAsyncOperations.rst:
```diff
@@ -0,0 +1,179 @@
+
+``async_load_to_lds``
+---------------------
+
```
```suggestion
Legacy LDS DMA (GFX9-10)
---------------------------------------
```

I wonder if this should be separated in two sections ? e.g.

* Synchronization Operations
  * asyncmark
  * wait.asyncmark(N)
* Memory Operations
  * Legacy LDS DMA (GFX9-10)

Pierre-vh commented on llvm/docs/AMDGPUAsyncOperations.rst:
```diff
@@ -0,0 +1,179 @@
+
+N is an unsigned integer; the ``N+1`` th predecessor marker of point X is a
+marker M such that there are `N` markers in program order from M to X, not
+including M.
```
If there are <N markers before M, how does it work ?
Is M just the oldest marker in that case ? I'd clarify it

ssahasra replied:
If there are less than N markers, then there is no M, and then the "if it exists" clause from the previous paragraph kicks in.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1207,7 +1261,56 @@ void WaitcntBrackets::print(raw_ostream &OS) const {
+      unsigned MarkedScore = Marker[T];
+      switch (T) {
+      case LOAD_CNT:
+        OS << "    " << (ST->hasExtendedWaitCounts() ? "LOAD" : "VM")
```
I feel like there is a good opportunity to simplify this and the switch above as well. e.g. move all of the printing part into a `toString(T)` util so this entire switch and the one above become `OS << toString(T) << ": " << MarkedScore` ?

ssahasra replied:
That was my initial thought too. But the format created is different for the two debug outputs. It's difficult to reconcile them into one.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2439,6 +2597,68 @@ bool WaitcntBrackets::mergeScore(const MergeInfo &M, unsigned &Score,
+  // Pad with zero-filled markers if our list is shorter.
+  // Zero represents "no pending async operations at this checkpoint"
+  // and acts as the identity element for max() during merging
+  std::array<unsigned, NUM_INST_CNTS> ZeroMarker{};
```
```suggestion
  std::array<unsigned, NUM_INST_CNTS> ZeroMarker;
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2517,6 +2740,11 @@ bool WaitcntBrackets::merge(const WaitcntBrackets &Other) {
+  StrictDom |= mergeAsyncMarkers(MergeInfos, Other.AsyncMarkers);
+  for (auto T : inst_counter_types(Context->MaxCounter)) {
+    StrictDom |= mergeScore(MergeInfos[T], AsyncScore[T], Other.AsyncScore[T]);
+  }
```
```suggestion
  for (auto T : inst_counter_types(Context->MaxCounter)) 
    StrictDom |= mergeScore(MergeInfos[T], AsyncScore[T], Other.AsyncScore[T]);
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1265,13 +1368,59 @@ void WaitcntBrackets::determineWaitForScore(InstCounterType T,
+    unsigned ScoreToWait = RequiredMarker[T];
+    if (ScoreToWait == 0) {
+      continue;
+    }
```
```suggestion
    if (ScoreToWait == 0) 
      continue;
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2439,6 +2597,68 @@ bool WaitcntBrackets::mergeScore(const MergeInfo &M, unsigned &Score,
 }
 
+bool WaitcntBrackets::mergeAsyncMarkers(
+    const MergeInfo MergeInfos[NUM_INST_CNTS],
```
tiny nit: passing an array by argument is technically fine but does feel a bit weird, how about `ArrayRef` instead ?

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIDefines.h:
```diff
@@ -368,12 +368,15 @@ enum CPol {
   DLC = 4,
+  SWZ_pregfx12 = 8,
   SCC = 16,
+  ASYNC_pregfx12 = 32,
```
Have you considered using a MMO flag instead ?

I think this ship has sailed, but I don't like that CPol has been turned into a half HW bits, half compiler information tracking utility.
If we'd like to keep using it this way we should really refactor it a bit at some point so the documentation is accurate, and there is a clear way to tell "fake" from "HW" bits apart.

krzysz00 replied:
I figure this should be an MMO flag at the isel level. At the IR level, smuggling this state into the high bits of CPol is a backwards-compatible way to have state that, unlike metadata, can't be dropped.

I'd be willing to clear the existing volatile out of CPol if we went and upgraded all the intrinsics to take an `immarg i1 %volatile` instead

krzysz00 replied:
Metadata wouldn't have worked because metadata needs to be erasable.

Now, I do think that if we want to do this properly, we make an `i1` immarg for whether an operation is async and autoupgrade in a value of `false`. I'd also say we may want to retroactively do that for `volatile` so that we can get non-hardware bits out of CPol. (I'm just not volunteering to write that code right now)

ssahasra replied:
I had not thought of autoupgrade. That sounds like a great idea. I have not relished the current abuse of CPol for this feature. Will try this out along with MMO.

krzysz00 replied:
... AutoIpgrade to add an argument to an intrinsic with a default value is a perfectly fine usage of that facility - we did it just recently for adding `clamp` to WMMA operations, for example 

And having buffer loads working is important, yes

ssahasra commented on llvm/docs/AMDGPUAsyncOperations.rst:
```diff
@@ -0,0 +1,179 @@
+       asyncmark();
+
+       wait.asyncmark(1); // will wait for at least the second block, possibly including bar()
+       wait.asyncmark(0); // will wait for third block, including bar()
```
Yes. I am not happy with the wording around function calls. The intention is that when foo calls bar, anything happening in bar is automatically covered but we can get inefficient waits. We get the same inefficient waits if bar got inlined, and it is still correct. See the (internal) discussions in SWDEV-521121 for how this was worked out.

krzysz00 commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -6594,12 +6594,18 @@ operations.
+
+- asynchronous: aux bit u0x40000000
+- volatile: aux bit u0x20
+- nontemporal: metadata
```
When I was adding volatile-in-Chpol, I specifically used a very high bit so that it would be clear that this is compiler-internal and not an actual hardware bit. If we're going to add another compiler-level flag like this, I strongly request that it sit very far away from hardware bits to avoid confusion, hence 1 << 30

krzysz00 commented:
Would the following statement be correct with respect to function calls?
> It is undefined behavior for a function (other than a program entry point) to include async mark operations without corresponding async wait operations, or for such a function to call async operations without calling asyncmark. Equivalently, no function may cause a net increase in the number of outstanding async operations. This allows us to assume that function calls cannot invalidate a local analysis of what hardware counter or counters must be waited on and what their values must decrease to in order to implement a given asyncwait

I know there's been a lot of fiddly points here, but I think this is the cleanest approach for something initial and we can relax that restriction later if needed.

krzysz00 commented:
Ok, let me write all this out just to make sure I've convinced myself of what you're going for in the hopes that it'll help improve the documentation.

One way to look at asyncwait(N) is that, when asyncwait(N), any operation that's within an async group other than one of the last N to be created must have completed. These async operations (ignoring the fact that the gfx9/10 ones key off of vmem) correspond to different hardware counters for different types of async operations.

What this implies is that, to materialize/implement asyncwait(N), the compiler must first determine all the sequences of async groups that must be the N alive ones. Then, for each such group and each counter type, we must determine a lower bound on the number of async operations that increment that counter that are in the group. For example, if we execute two async loads to LDS and one tensor load in the previous async group, we'd know we need asynccnt <= 2 and tensorcnt <= 1 after an asyncwait(1) - or that there are >= 2 async ops and >= 1 tensor ops in the previous 1 groups. In a more complex scenario like a double-buffered loop, we could have something like this

```
tensor_load(... to A1)
tensor_load(... to B1)
asyncmark()
for (..) {
  barrier
  tensor_load(... to A2)
  tensor_load(... to B2)
  asyncmark()
  asyncwait(1)
  barrier
  compute(A1, B1) // no async ops
  barrier
  tensor_load(... to A1)
  tensor_load(... to B1)
  asyncmark()
  asyncwait(1)
  barrier
  compute(A2, B2)
[
asyncwait(0)
barrier
compute(A1, B1)
```
we know that every one of those asyncwait(1) calls should be implemented with a tensor count wait of 2 and an async count wait of 0 that can be eliminated (that is, we have lower bounds of async count >= 0  and tensor count >= 2 if the only async operations com from the previously marked group ... and then a later analysis can go "hey wait a minute, there's an async wait that's trivially a nop, let's delete it").

For another case,
```
tensor_load(...)
asyncmark() // A
async_to_lds(...)
asyncmark() // B
if (...) {
  async_to_lds(...)
  asyncmark() // C
} else {
  tensor_load()
  asyncmark() // D
}
asyncwait(2)
```
the asyncwait will become asynccnt_wait(1) and tensorcnt_wait(0) because
- If the two previous marks are C and B, we'll need two async waits outstanding and no tensor waits outstanding
- If the two previous groups are D and B, we'll need one asyncwait and one tensor wait outstanding

Intersecting these two bounds gives us a pattern of waits that'll always work, and, unlike in the previous example, we can't eliminate that wait to 0 tensor ops, because it's possible that the program execution path was A B C, which means we need the wait operation to drive he tensor load from A to completion.

Now, function calls.

Since the thing we care about is lower bounds for how much each mark increments each async counter, it's safe to assume that each function call does not increment this counter. That's the key point I was missing until now. If the function _did_ increment the counter, we'd just be driving operations that we are explicitly looking at in our analysis of its asyncwait()-using caller to completion because we didn't know about the new ones the function issued.

Now, if we can that the number of async operations of a particular kind a function launches is in [L, U] for some L and U, we can be less conservative. In particular, if we know that some function f does not contain any async operations of a given kind (or, most generally, if we could stick an `amdgpu-noasync` on it), then spurious async waits (say, a tensor-count wait in a program that never does any tensor loads) could be eliminated by the logic that does that sort of thing. If we had more precise knowledge what the function was doing - let's say we know it created no async calls and between 2 and 4 tensor DMAs, we'd be able to use that information to get more precise upper bounds on our asyncwait()s, but that might be out of scope for initial work.

Now, suppose that there's an asyncwait(N) inside a non-entry-point function. If that function established at least N async marks, then everything's fine and we just analyze those locally. The callers of that function don't need to know about this - it just means that their own local analysis of how the marks work will insert waits for operations that have already been driven to completion. (the most extreme example being a function that starts or ends with an asyncwai(0) and never does any async operations, which will make the caller's more precise analysis pointless by way of introducing the "memory is synchronized at "calls" we have for non-async ops[.

However, if a function issues N async marks and than does an asyncwait(N + K[, we don't know anything abut that function's callers (ok maybe sometimes we do but we're not going there in v1 or possibly ever) and so it can only formulate its lower bounds based on the N async marks it has access to. So this'll also be safe, but it won't be performant or do what one really expects.

What all this means is that, while we probably want an attributor scheme for performance's sake, we don't need to reason about the contents of called functions for correctness like you said and it took me a while to realize.

There's also an interesting corollary here: it is always correct to replace asyncwait(N) with asyncwait(M) where M < N - it'll be bad for performance, but it'll be correct.

krzysz00 commented:
@ssahasra Does the stuff I wrote above match your understanding of what we're going for, and will it help update the documentation?

arsenm commented on clang/include/clang/Basic/BuiltinsAMDGPU.def:
```diff
@@ -538,6 +538,14 @@ TARGET_BUILTIN(__builtin_amdgcn_ashr_pk_u8_i32, "UsUiUiUi", "nc", "ashr-pk-insts
+//===----------------------------------------------------------------------===//
+
+// FIXME: Not supported on GFX12 yet. Will need a new feature when we do.
+TARGET_BUILTIN(__builtin_amdgcn_asyncmark, "v", "n", "vmem-to-lds-load-insts")
```
This needs to be updated for the move to tablegen

krzysz00 reviewed (commented):
Missing test: a combination of async DMA ops and regular `load` operations from global on gfx9

Missing test: a pipelined loop example wheer we show that you don't get conservative waitcnt insertion when using the async intrinsics

krzysz00 reviewed (commented):
Also, AMDGPULowerBufferFatPointers should be updated for the new load.to.lds intrinsic

ssahasra commented:
> Missing test: a combination of async DMA ops and regular `load` operations from global on gfx9
> 
> Missing test: a pipelined loop example wheer we show that you don't get conservative waitcnt insertion when using the async intrinsics

I am not sure that I follow. The new test `asyncmark-pregfx12.ll` seems to contain exactly that and a bit more? The LLVM IR even has comments explaining the effect of the "regular" global loads on vmcnt.

ssahasra commented on clang/include/clang/Basic/BuiltinsAMDGPU.td:
```diff
@@ -288,7 +298,9 @@ def __builtin_amdgcn_global_atomic_fadd_v2bf16 : AMDGPUBuiltin<"_ExtVector<2, sh
 def __builtin_amdgcn_load_to_lds : AMDGPUBuiltin<"void(void *, void address_space<3> *, _Constant unsigned int, _Constant int, _Constant unsigned int)", [], "vmem-to-lds-load-insts">;
+def __builtin_amdgcn_load_async_to_lds : AMDGPUBuiltin<"void(void *, void address_space<3> *, _Constant unsigned int, _Constant int, _Constant unsigned int)", [], "vmem-to-lds-load-insts">;
 def __builtin_amdgcn_global_load_lds : AMDGPUBuiltin<"void(void address_space<1> *, void address_space<3> *, _Constant unsigned int, _Constant int, _Constant unsigned int)", [], "vmem-to-lds-load-insts">;
+def __builtin_amdgcn_global_load_async_lds : AMDGPUBuiltin<"void(void address_space<1> *, void address_space<3> *, _Constant unsigned int, _Constant int, _Constant unsigned int)", [], "vmem-to-lds-load-insts">;
```
I think I'll leave it in just for the sake of completion. Unless you are anticipating some future maintenance cost that we could avoid?

krzysz00 reviewed (commented):
@ssahasra Buffer fat pointer changes lgtm

ssahasra commented:
> > Missing test: a pipelined loop example wheer we show that you don't get conservative waitcnt insertion when using the async intrinsics
> 
> I am not sure that I follow. The new test `asyncmark-pregfx12.ll` seems to contain exactly that and a bit more? The LLVM IR even has comments explaining the effect of the "regular" global loads on vmcnt.

The loop test is now improved to run with and without async LDS ops. When the resulting CHECK lines are placed side-by-side, we can see how the vmcnt generated is different.

ssahasra commented:
All checks passed, and no recent merge conflicts. I need an LGTM (preferably two) to submit this. @nhaehnle @arsenm @jayfoad

Pierre-vh reviewed (commented):
I wish this was split into multiple patches because it's quite hard to review all at once. Not sure how other reviewers feel about it.

Can you update the commit description to give a more accurate overview of all the contents of this patch?

ssahasra committed:
[AMDGPU] Introduce asyncmark/wait intrinsics

Asynchronous operations are memory transfers (usually between the global memory
and LDS) that are completed independently at an unspecified scope. A thread that
requests one or more asynchronous transfers can use *async marks* to track their
completion. The thread waits for each mark to be *completed*, which indicates
that requests initiated in program order before this mark have also completed.

For now, we implement asyncmark/wait operations on pre-GFX12 architectures that
support "LDS DMA" operations. Future work will extend support to GFX12Plus
architectures that support "true" async operations.

Co-authored-by: Ryan Mitchell <ryan.mitchell@amd.com>

Fixes: SWDEV-521121
